extensions/amp-form/0.1/test/integration/test-integration-form.js

```diff
@@ -0,0 +1,229 @@
+    }
+
+    if (config.errorTemplate) {
+      const successDiv = doc.createElement('div');
+      successDiv.setAttribute('submit-error', '');
+      const template = doc.createElement('template');
+      template.setAttribute('type', 'amp-mustache');
+      template./*OK*/innerHTML = config.errorTemplate;
+      successDiv.appendChild(template);
+      form.appendChild(successDiv);
+    }
+
+    doc.body.appendChild(form);
+    return form;
+  }
+
+  describe('on=submit:form.submit', () => {
+    it('should be protected from recursive-submission', () => {
+      const form = getForm({
+        id: 'sameform',
+        actionXhr: baseUrl + '/form/post',
+        on: 'submit:sameform.submit',
+      });
+      const ampForm = new AmpForm(form);
+      sandbox.spy(ampForm, 'handleXhrSubmit_');
+      sandbox.spy(ampForm, 'handleSubmitAction_');
+      sandbox.spy(ampForm.xhr_, 'fetch');
+      form.dispatchEvent(new Event('submit'));
+      return timer.promise(10).then(() => {
+        // Due to recursive nature of 'on=submit:sameform.submit' we expect
+        // the action handler to be called twice, the first time for the
+        // actual user submission.
+        // The second time in response to the `submit` event being triggered
+        // and sameform.submit being invoked.
+        expect(ampForm.handleSubmitAction_).to.have.been.calledTwice;
+
+        // However, only the first invocation should be handled completely.
+        // and any subsequent calls should be stopped early-on.
+        expect(ampForm.handleXhrSubmit_).to.have.been.calledOnce;
+        expect(ampForm.xhr_.fetch).to.have.been.calledOnce;
+      });
+    });
+  });
+
+  describe('Submit xhr-POST', () => {
+    it('should submit and render success', () => {
+      const form = getForm({
+        id: 'form1',
+        actionXhr: baseUrl + '/form/post/success',
+        successTemplate: 'Thanks {{name}} for adding your interests:' +
+            ' {{#interests}}{{title}} {{/interests}}.',
+        errorTemplate: 'Should not render this.',
+      });
+      new AmpForm(form);
+      form.dispatchEvent(new Event('submit'));
+      return timer.promise(100).then(() => {
+        const rendered = form.querySelectorAll('[i-amp-rendered]');
+        expect(rendered.length).to.equal(1);
+        expect(rendered[0].textContent).to.equal(
+            'Thanks John Miller for adding your interests: ' +
+            'Football Basketball Writing .');
+      });
+    });
+
+    it('should submit and render error', () => {
+      const form = getForm({
+        id: 'form1',
+        actionXhr: baseUrl + '/form/post/error',
+        successTemplate: 'Should not render this.',
+        errorTemplate: 'Oops. {{name}} your email {{email}} is already ' +
+            'subscribed.',
+      });
+      new AmpForm(form);
+      const errors = [];
+      const realSetTimeout = window.setTimeout;
+      sandbox.stub(window, 'setTimeout', (callback, delay) => {
```

Comment:
Ah yeah, I am doing this because of the `rethrowAsync` call inside the catch callback of the fetch call. The only way I could think of catching the async error is through stubbing timeout and forcing a try-catch clause to avoid the tests flaking because of async-thrown error.